velox/functions/prestosql/FromUtf8.cpp

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/StringWriter.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/string/StringImpl.h"
+#include "velox/functions/prestosql/Utf8Utils.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+class FromUtf8Function : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    const auto& input = args[0];
+
+    // Optimize common cases: (1) all-ASCII input; (2) valid UTF-8 input.
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto& decodedInput = *decodedArgs.at(0);
+    if (input->as<SimpleVector<StringView>>()->computeAndSetIsAscii(rows)) {
+      // Input strings are all-ASCII.
+      toVarcharNoCopy(input, decodedInput, rows, context, result);
+      return;
+    }
+
+    auto firstInvalidRow = findFirstInvalidRow(decodedInput, rows);
+    if (!firstInvalidRow.has_value()) {
+      // All inputs are valid UTF-8 strings.
+      toVarcharNoCopy(input, decodedInput, rows, context, result);
+      return;
+    }
+
+    BaseVector::ensureWritable(rows, VARCHAR(), context.pool(), result);
+    auto flatResult = result->as<FlatVector<StringView>>();
+
+    // Reserve string buffer capacity.
+    size_t totalInputSize = 0;
+    rows.applyToSelected([&](auto row) {
+      totalInputSize += decodedInput.valueAt<StringView>(row).size();
+    });
+
+    flatResult->getBufferWithSpace(totalInputSize);
+
+    // Optimize for common case of constant 'replacement' argument.
+    bool constantReplacement =
+        args.size() == 1 || decodedArgs.at(1)->isConstantMapping();
+
+    if (constantReplacement) {
+      std::string replacement = kReplacementChar;
+      if (args.size() > 1 && decodedArgs.at(1)->isConstantMapping()) {
+        try {
+          replacement =
+              getReplacementCharacter(args[1]->type(), *decodedArgs.at(1), 0);
+        } catch (const std::exception&) {
+          context.setErrors(rows, std::current_exception());
```

Comment:
nit: maybe just just give a name to exception and:

```
context.setErrors(rows, e);
```

Reply:
@pedroerp setErrors requires std::exception_ptr, so I cannot do what you suggested.

```
void EvalCtx::setErrors(
    const SelectivityVector& rows,
    const std::exception_ptr& exceptionPtr) {
```